handler/src/main/java/io/netty/handler/timeout/ReadAfterWriteTimeoutHandler.java

```diff
@@ -0,0 +1,221 @@
+    public ReadAfterWriteTimeoutHandler(long timeout, TimeUnit unit) {
+        this(timeout, unit, true, true);
+    }
+
+    @Override
+    public void handlerAdded(final ChannelHandlerContext ctx) throws Exception {
+        removed = false;
+
+        if (channelOnWriteListener == null) {
+            channelOnWriteListener = new ChannelFutureListener() {
+                @Override
+                public void operationComplete(ChannelFuture future) throws Exception {
+                    //ensure that only the last write listener in a batch would setup one final timeout task
+                    //or nothing in a case of error
+                    disable();
+                    timedOut = false;
+                    if (future.isSuccess()) {
+                        if (removed) {
+                            return;
+                        }
+                        writeFinishTime = System.nanoTime();
+                        //init readFinishTime to writeFinishTime - readFinishTime > 0 be true if we don't get response
+                        readFinishTime = writeFinishTime - 1;
+                        if (timeoutNanos > 0) {
+                            timeout = ctx.executor().schedule(readTimeoutTask, timeoutNanos, TimeUnit.NANOSECONDS);
+                        }
+                    }
+                }
+            };
+        }
+
+        if (readTimeoutTask == null) {
+            readTimeoutTask = new ReadTimeoutTask(ctx);
+        }
+    }
+
+    @Override
+    public void handlerRemoved(ChannelHandlerContext ctx) throws Exception {
+        destroy();
+    }
+
+    @Override
+    public void channelInactive(ChannelHandlerContext ctx) throws Exception {
+        destroy();
+        super.channelInactive(ctx);
+    }
+
+    @Override
+    public void channelRead(ChannelHandlerContext ctx, Object msg) throws Exception {
+        readFinishTime = System.nanoTime();
+        if (timeoutShouldBeCancelled(ctx, msg)) {
+            disable();
+        }
+        if (suppressReadAfterTimeout && timedOut) {
+            ctx.fireUserEventTriggered(ReadSuppressed.INSTANCE);
+            ReferenceCountUtil.release(msg);
+        } else {
+            super.channelRead(ctx, msg);
+        }
+    }
+
+    @Override
+    public void write(ChannelHandlerContext ctx, Object msg, ChannelPromise promise) {
+        ctx.write(msg, promise.addListener(channelOnWriteListener));
+    }
+
+    /**
+     * This method is called after receiving each {@link #channelRead} and should define if this read should trigger
+     * timeout cancelling. Default implementation returns {@code true}.
+     * @param ctx {@link ChannelHandlerContext} of related {@link #channelRead} invocation
+     * @param msg message of related {@link #channelRead} invocation
+     * @return true if timeout should be cancelled by receiving this read, false otherwise
+     */
+    protected boolean timeoutShouldBeCancelled(@SuppressWarnings("unused") ChannelHandlerContext ctx,
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Remove this unused method parameter "ctx". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://garage.netty.io/sonarqube/coding_rules#rule_key=squid%3AS1172)